### ISSUE TITLE
Render Lidar Points on Sweep images

### ISSUE THREAD

**Issue by cathrinasowa** [nutonomy/nuscenes-devkit#160]
Hi,
I would like to know if it is possible to render Lidar points to an image which is in a sweep but not in a sample. I know that the Lidar and camera are aligned for samples only. Nevertheless would it be possible to render the Lidar pointcloud with the nearest timestamp to the given image.
Anyhow is it possible to loop over images in a sweep by a token or do they have a token that can be used?
In the data colection description it is said that not every Lidar Scan has a corresponding camera frame but does every camera frame has a Lidar scan?
I hope you can help me as quick and helpfull as the other time I asked a question.
Thank you in advance for any help.

Greetings
Cathrina

**Comment by holger-nutonomy** [comment]
Hi Cathrina,
it is definitely "possible" to do that. In fact I was once rendering videos with the lidar points projected onto the camera view (at around 12Hz = camera frequency) and did not see any issues. [map_pointcloud_to_image()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L515) works with an arbitrary pointcloud and camera token. So you can either write your own function or you just modify (a copy of) [render_pointcloud_in_image()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L576). E.g. to take the next camera image instead, you can do:
```
  camera_token = nusc.get('sample_data', sample_record['data'][camera_channel])['next']
```
To "loop over images in a sweep" you can start from the previous sample, then for every camera go to the next camera image until you are "close" to the timestamp of the lidar sweep. Then store that in a list, which you can later loop over. 
To give you a related example look at [render_scene()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L936) which creates a movie of the scene. At every timestep it advances by a fixed time unit, then for every sensor advances the pointer to the sample_data just after the timestep. Your example is a bit trickier as the "closest" image could be before or after the lidar sweep.

> In the data collection description it is said that not every Lidar Scan has a corresponding camera frame but does every camera frame have a Lidar scan?

Basically yes. Although the word "corresponding" is itself vague (our fault). Note that a lidar sweep takes 1/20s, but taking a camera image is near-instantaneous. So the camera is triggered when the lidar sweeps across the center of its FOV, but some of the points are newer, some are older. The timestamp  "of the LIDAR scan is the time when the full rotation of the current LIDAR frame is achieved".

Please let me know if you have any more questions.

**Comment by cathrinasowa** [comment]
Hey Holger,
Thank again for your fast,detailed and helpful answer. I missed the 'next' option and was confused on how to get the images of the sweeps. The images of a sample can be reached by the smaple tokens and I thought that there have to be something similar for the sweeps.
I wrote my own method and now I am able  to project the Lidar points to the image I chose.
 Have a nice day
Cathrina

**Comment by Shaoqing26** [comment]
> Hey Holger,
> Thank again for your fast,detailed and helpful answer. I missed the 'next' option and was confused on how to get the images of the sweeps. The images of a sample can be reached by the smaple tokens and I thought that there have to be something similar for the sweeps.
> I wrote my own method and now I am able to project the Lidar points to the image I chose.
> Have a nice day
> Cathrina

Hi Cathrina,
     I also have problem in dealing this question,I go though each sample,and every sample has nine sweeps,First i project sample lidar point on each camera, When finished it ,use **sweep_lidar_token = nusc.get('sample_data',pointsensor_token)['prev']** to get previous sweep,because sweep do not have token,so i cant find related cameras of sweep,so i save sample each camera token,total six, Then i use **camera_token = nusc.get('sample_data',_sample_token)['prev']** to get current sweeps token,However ,It doest work,after a few sample,the program was breakup,The camera token get ' ',so it cant get any data,I think the problems is that sweep dont like sample each lidar pointcloud has related one camera,so,lidarpoint dont have itself camera, I wana to fix it,Can u share your code?  OR  appreciate your kindly  advice.

**Comment by Shaoqing26** [comment]
> Hi Cathrina,
> it is definitely "possible" to do that. In fact I was once rendering videos with the lidar points projected onto the camera view (at around 12Hz = camera frequency) and did not see any issues. [map_pointcloud_to_image()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L515) works with an arbitrary pointcloud and camera token. So you can either write your own function or you just modify (a copy of) [render_pointcloud_in_image()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L576). E.g. to take the next camera image instead, you can do:
> 
> ```
>   camera_token = nusc.get('sample_data', sample_record['data'][camera_channel])['next']
> ```
> 
> To "loop over images in a sweep" you can start from the previous sample, then for every camera go to the next camera image until you are "close" to the timestamp of the lidar sweep. Then store that in a list, which you can later loop over.
> To give you a related example look at [render_scene()](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/nuscenes.py#L936) which creates a movie of the scene. At every timestep it advances by a fixed time unit, then for every sensor advances the pointer to the sample_data just after the timestep. Your example is a bit trickier as the "closest" image could be before or after the lidar sweep.
> 
> > In the data collection description it is said that not every Lidar Scan has a corresponding camera frame but does every camera frame have a Lidar scan?
> 
> Basically yes. Although the word "corresponding" is itself vague (our fault). Note that a lidar sweep takes 1/20s, but taking a camera image is near-instantaneous. So the camera is triggered when the lidar sweeps across the center of its FOV, but some of the points are newer, some are older. The timestamp "of the LIDAR scan is the time when the full rotation of the current LIDAR frame is achieved".
> 
> Please let me know if you have any more questions.

I try your advice, but has problems,i mentioned above,can u help me find out where the question is ?
thanks~~~~
[point2Nusc_img .txt](https://github.com/nutonomy/nuscenes-devkit/files/6027447/point2Nusc_img.txt)

**Comment by holger-nutonomy** [comment]
@Shaoqing26 When `prev` equals `""`, it means that we are at the beginning of that scene, hence there is no previous data.

**Comment by Shaoqing26** [comment]
> @Shaoqing26 When `prev` equals `""`, it means that we are at the beginning of that scene, hence there is no previous data.

yes,I know that,each sample‘s **prev** of beginning of every scenes is **" "**,I had judge it，if the first sample of every scenes , I will only project lidar to six camera， it wont go through to get sweeps,but is isnt,it must be to go through 9 sweeps before that sample,the problems is that there has no sweeps token ,so wo cant project lidar to cameras like samples data, i go through it by sample' camera token['prev] to get closest sweep,and do it agin to get sweep closest that sweep, is there any problems? I get error when  through several times of that，there has no camera token .....

**Comment by holger-motional** [comment]
Only for a sample do we guarantee that lidar and camera are synchronized (close in time). For other camera-lidar pairs (sweeps) you can try, but they may be taken at significantly different moments in time and therefore the objects may have moved between the image and pointcloud.
